python/tvm/build_module.py

```diff
@@ -52,17 +57,58 @@ def __getattr__(self, name):
             return BuildConfig.defaults[name]
         return self._attr[name]
 
+    def decorator(self, func):
+        ''' decorate the pass function'''
+        def dump(*args, **kwargs):
+            '''dump function'''
+            retv = func(*args, **kwargs)
+            # pylint: disable=unidiomatic-typecheck
+            if type(retv) not in self._types:
+                return retv
+            pname = str(self._pass_id) + "_" + func.func_name + "_ir.cc"
+            with open(pname, "a") as f:
+                out = retv.body if isinstance(retv, container.LoweredFunc) else retv
+                print(out, file=f)
+                if isinstance(retv, container.Array):
+                    for x in retv:
+                        out = x.body if isinstance(x, container.LoweredFunc) else x
+                        print("---------", x.name, "\n", out, "\n-----------\n", file=f)
+                self._pass_id += 1
+            return retv
+        return dump
+
+    def decorate_irpass(self):
+        for k, v in vars(ir_pass).items():
+            vars(ir_pass)[k] = self.decorator(v) if isinstance(v, types.FunctionType) else v
+        schedule.ScheduleOps = self.decorator(schedule.ScheduleOps)
+        self._types = [v for k, v in vars(_stmt).items()]
+        self._types.append(container.Array)
+        self._types.append(container.LoweredFunc)
+
+    def decorate_custompass(self, add_lower_pass):
+        if self.dump_pass_ir is False:
+            return add_lower_pass
+        self._pass_id = 0
```

Comment:
The update rule of pass_id is mysterious. 
The code logic always becomes harder to understand when you have states that have side-effect among function calls, pass_id is such stage. I would rather.

If you want event counting, name it to pass_counter. Only have reset in one place(when you enter) and increment one place(when you dump)

Reply:
the case is if you call multiple lower, 
def test1():
      func1
      lower
def test2():
      func2
      lower

with build cfg(dump_pass_ir=true)
  test1()
  test2()

if we don't reset pass_id, you will get two batchs of dump file, like 0_pass1_ir.cc, ... 16_pass17_ir.cc, 17_pass1_ir.cc, ... 33_pass17_ir.cc, 
after reset, we can fuse them as one batch with both func1 and func2 in it